packages/viewer/src/lib/walk-mode/interactions.ts

```diff
@@ -0,0 +1,208 @@
+
+    this.pressed = Object.keys(this.pressed)
+      .filter((k) => k !== key)
+      .reduce((pressed, k) => ({ ...pressed, [k]: true }), {});
+
+    if (Object.keys(this.pressed).length === 0 && this.interval != null) {
+      this.endInteraction();
+    }
+  };
+
+  private handleEnabledChange(enabled: boolean): void {
+    if (enabled) {
+      this.enable();
+    } else {
+      this.disable();
+    }
+  }
+
+  private restartInteraction(): void {
+    this.clearInterval();
+    this.tryBeginInteraction();
+  }
+
+  private tryBeginInteraction(): void {
+    if (Object.keys(this.pressed).length > 0 && this.someOperationMatches()) {
+      this.beginInteraction();
+    }
+  }
+
+  private beginInteraction(): void {
+    if (!this.api?.isInteracting()) {
+      this.api?.beginInteraction();
+    }
+
+    if (this.interval == null) {
+      this.interval = setInterval(
+        this.updateCamera,
+        this.model.getKeyboardRepeatInterval()
+      );
+    }
+  }
+
+  private endInteraction = async (): Promise<void> => {
+    this.clearInterval();
+    await this.api?.endInteraction();
+  };
+
+  private clearInterval(): void {
+    if (this.interval != null) {
+      clearInterval(this.interval);
+      this.interval = undefined;
+    }
+  }
+
+  private updateCamera(): void {
+    Object.keys(this.handlers).forEach((key) => {
+      const op = key as ViewerWalkModeOperation;
+
+      if (this.model.operationMatches(op, this.pressed)) {
+        this.handlers[op]();
+      }
+    });
+  }
+
+  private someOperationMatches(): boolean {
+    return Object.keys(this.handlers).some((op) =>
+      this.model.operationMatches(op as ViewerWalkModeOperation, this.pressed)
+    );
+  }
+
+  private pivotLeft(): void {
+    this.api?.pivotCamera(0, this.model.getKeyboardPivotDegrees());
+  }
+
+  private pivotRight(): void {
+    this.api?.pivotCamera(0, -this.model.getKeyboardPivotDegrees());
+  }
+
+  private pivotUp(): void {
+    this.api?.pivotCamera(-this.model.getKeyboardPivotDegrees(), 0);
+  }
+
+  private pivotDown(): void {
+    this.api?.pivotCamera(this.model.getKeyboardPivotDegrees(), 0);
+  }
+
+  private walkForward(): void {
+    this.api?.walk(Vector3.scale(this.relativeWalkSpeed(), Vector3.forward()));
+  }
+
+  private walkBackward(): void {
+    this.api?.walk(Vector3.scale(this.relativeWalkSpeed(), Vector3.back()));
+  }
+
+  private walkLeft(): void {
+    this.api?.walk(Vector3.scale(this.relativeWalkSpeed(), Vector3.left()));
+  }
+
+  private walkRight(): void {
+    this.api?.walk(Vector3.scale(this.relativeWalkSpeed(), Vector3.right()));
+  }
+
+  private moveUp(): void {
+    this.api?.walk(Vector3.scale(this.relativeWalkSpeed(), Vector3.down()));
+  }
+
+  private moveDown(): void {
+    this.api?.walk(Vector3.scale(this.relativeWalkSpeed(), Vector3.up()));
+  }
+
+  private relativeWalkSpeed(): number {
+    const speedScalar = (5 / this.model.getKeyboardWalkSpeed()) * 100;
+    return 1 / speedScalar;
```

Comment:
Is there an advantage to writing it this way as opposed to `return this.model.getKeyboardWalkSpeed()/500`?

Reply:
Yeah, this is probably overly confusing. This was pulled directly from Connect, and I'm guessing there used to be some additional logic here. I'll simplify this a bit!